### ISSUE TITLE
Should check if self._timeout_heap is None before pushing on heap

### ISSUE THREAD

**Issue by prot0man** [pika/pika#1291]
https://github.com/pika/pika/blob/a3ae3e5aae3fd546e9ae73bff0175f5d61a9d39a/pika/adapters/select_connection.py#L258
The first argument to heaq.heappush must be a list. If the close is called on the Timer object, then self._timeout_heap will be set to None, which will cause exceptions with the heappush. Adding a check for "if self.timeout_heap is not None" at the beginning of the function should fix the issue.

**Comment by lukebakken** [comment]
Thanks for noticing this. We have not received any reports of this causing an issue. If you'd like, submit a pull request to address this issue.

**Comment by prot0man** [comment]
So I've seen this line of code cause the exception "TypeError: heap argument must be a list", and that's how I found the issue. I'll try to get a pull request in.

**Comment by lukebakken** [comment]
It would be great to know how to reproduce the `TypeError` exception.

**Comment by prot0man** [comment]
Honestly, I'm trying to make it reproduce myself. I had a code base that was using pika working for the longest time without seeing the issue, then when I installed that codebase on a new VM, I ran into this issue. I still can't reproduce the issue on my main machine and don't have access to the VM at the moment.

**Comment by prot0man** [comment]
All I know for now is that self._timeout_heap is None when that callback is being called.

**Comment by mcguiremw** [comment]
I am able to get this error when calling `self.connection.process_data_events(time_limit=0)`

```> File "/app/processor.py", line 1017, in _update
    self.connection.process_data_events(time_limit=0)
    │    │          └ <function BlockingConnection.process_data_events at 0x7ff05f498550>
    │    └ <BlockingConnection impl=<SelectConnection CLOSED transport=None params=<ConnectionParameters host=rabbitmq. port=5672 virtua...
    └ <__main__.MessageBuffer object at 0x7ff06c618790>

  File "/usr/local/lib/python3.9/site-packages/pika/adapters/blocking_connection.py", line 827, in process_data_events
    with _IoloopTimerContext(time_limit, self._impl) as timer:
         │                   │           │    └ <SelectConnection CLOSED transport=None params=<ConnectionParameters host=rabbitmq. port=5672 virtual_host=/ ssl=False>>
         │                   │           └ <BlockingConnection impl=<SelectConnection CLOSED transport=None params=<ConnectionParameters host=rabbitmq. port=5672 virtua...
         │                   └ 0
         └ <class 'pika.adapters.blocking_connection._IoloopTimerContext'>
  File "/usr/local/lib/python3.9/site-packages/pika/adapters/blocking_connection.py", line 189, in __enter__
    self._timer_handle = self._connection._adapter_call_later(
    │    │               │    │           └ <function BaseConnection._adapter_call_later at 0x7ff05f4b11f0>
    │    │               │    └ <SelectConnection CLOSED transport=None params=<ConnectionParameters host=rabbitmq. port=5672 virtual_host=/ ssl=False>>
    │    │               └ <pika.adapters.blocking_connection._IoloopTimerContext object at 0x7ff05eea03d0>
    │    └ None
    └ <pika.adapters.blocking_connection._IoloopTimerContext object at 0x7ff05eea03d0>
  File "/usr/local/lib/python3.9/site-packages/pika/adapters/base_connection.py", line 205, in _adapter_call_later
    return self._nbio.call_later(delay, callback)
           │    │     │          │      └ <bound method _CallbackResult.signal_once of <pika.adapters.blocking_connection._CallbackResult object at 0x7ff06060c380>>
           │    │     │          └ 0
           │    │     └ <function SelectorIOServicesAdapter.call_later at 0x7ff05f48cdc0>
           │    └ <pika.adapters.utils.selector_ioloop_adapter.SelectorIOServicesAdapter object at 0x7ff05ee7d670>
           └ <SelectConnection CLOSED transport=None params=<ConnectionParameters host=rabbitmq. port=5672 virtual_host=/ ssl=False>>
  File "/usr/local/lib/python3.9/site-packages/pika/adapters/utils/selector_ioloop_adapter.py", line 222, in call_later
    return _TimerHandle(self._loop.call_later(delay, callback), self._loop)
           │            │    │     │          │      │          │    └ <pika.adapters.select_connection.IOLoop object at 0x7ff05ee7d8e0>
           │            │    │     │          │      │          └ <pika.adapters.utils.selector_ioloop_adapter.SelectorIOServicesAdapter object at 0x7ff05ee7d670>
           │            │    │     │          │      └ <bound method _CallbackResult.signal_once of <pika.adapters.blocking_connection._CallbackResult object at 0x7ff06060c380>>
           │            │    │     │          └ 0
           │            │    │     └ <function IOLoop.call_later at 0x7ff05f491430>
           │            │    └ <pika.adapters.select_connection.IOLoop object at 0x7ff05ee7d8e0>
           │            └ <pika.adapters.utils.selector_ioloop_adapter.SelectorIOServicesAdapter object at 0x7ff05ee7d670>
           └ <class 'pika.adapters.utils.selector_ioloop_adapter._TimerHandle'>
  File "/usr/local/lib/python3.9/site-packages/pika/adapters/select_connection.py", line 444, in call_later
    return self._timer.call_later(delay, callback)
           │    │      │          │      └ <bound method _CallbackResult.signal_once of <pika.adapters.blocking_connection._CallbackResult object at 0x7ff06060c380>>
           │    │      │          └ 0
           │    │      └ <function _Timer.call_later at 0x7ff05f491040>
           │    └ <pika.adapters.select_connection._Timer object at 0x7ff05ee7d7f0>
           └ <pika.adapters.select_connection.IOLoop object at 0x7ff05ee7d8e0>
  File "/usr/local/lib/python3.9/site-packages/pika/adapters/select_connection.py", line 258, in call_later
    heapq.heappush(self._timeout_heap, timeout)
    │     │        │    │              └ <pika.adapters.select_connection._Timeout object at 0x7ff05eea03a0>
    │     │        │    └ None
    │     │        └ <pika.adapters.select_connection._Timer object at 0x7ff05ee7d7f0>
    │     └ <built-in function heappush>
    └ <module 'heapq' from '/usr/local/lib/python3.9/heapq.py'>

TypeError: heap argument must be a list```

**Comment by lukebakken** [comment]
Fixed by #1292
